pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -2615,8 +2600,147 @@ protected PersistentOfflineTopicStats internalGetBacklog(boolean authoritative)
+                        clientAppId(), topicName, wae);
+            }
+            resumeAsyncResponseExceptionally(asyncResponse, wae);
+        } catch (Exception e) {
+            log.error("[{}] Failed to get backlog size for non-partitioned topic {}", clientAppId(), topicName, e);
+            resumeAsyncResponseExceptionally(asyncResponse, e);
+        }
+    }
+
+    private boolean isLedgerExists(MessageIdImpl messageId){
+        try {
+            String internalInfo = pulsar().getAdminClient().topics().getInternalInfo(topicName.toString());
+            boolean ledgerExists = false;
+            if(messageId.getLedgerId() == -1) {
+                ledgerExists = true;
+            } else {
+                if (topicName.isPartitioned()) {
+                    ManagedLedgerInfo managedLedgerInfo = jsonMapper().readValue(internalInfo, ManagedLedgerInfo.class);
+                    ledgerExists = managedLedgerInfo.ledgers.stream().anyMatch(ledgerInfo -> messageId.getLedgerId() == ledgerInfo.ledgerId);
+                } else {
+                    JsonNode partitionedInternalInfos = jsonMapper().readTree(internalInfo).get("partitions");
+                    for (Iterator<Map.Entry<String, JsonNode>> it = partitionedInternalInfos.fields(); it.hasNext(); ) {
+                        String partitionedInternalInfo = it.next().getValue().toString();
+                        ManagedLedgerInfo managedLedgerInfo = jsonMapper().readValue(partitionedInternalInfo, ManagedLedgerInfo.class);
+                        ledgerExists = ledgerExists || managedLedgerInfo.ledgers.stream().anyMatch(ledgerInfo -> messageId.getLedgerId() == ledgerInfo.ledgerId);
+                    }
+                }
+            }
+            return ledgerExists;
+        }  catch (Exception e) {
+            log.error("[{}] Failed to get managed info for {}", clientAppId(), topicName, e);
+            throw new RestException(e);
+        }
+    }
+
+    protected void internalGetBacklogSizeByMessageId(AsyncResponse asyncResponse, MessageIdImpl messageId, boolean authoritative, boolean checkLedgerExists) {
+        if (topicName.isGlobal()) {
+            try {
+                validateGlobalNamespaceOwnership(namespaceName);
+            } catch (Exception e) {
+                log.error("[{}] Failed to get backlog size for {}", clientAppId(), topicName, e);
+                resumeAsyncResponseExceptionally(asyncResponse, e);
+                return;
+            }
+        }
+
+        if(checkLedgerExists) {
+            boolean ledgerExists = isLedgerExists(messageId);
+            if (!ledgerExists) {
+                log.warn("[{}] ledger {} not exists on topic {}", messageId.getLedgerId(), clientAppId(), topicName);
+                asyncResponse.resume(new RestException(Status.NOT_FOUND, String.format("ledger %s not exists", messageId.getLedgerId())));
+                return;
+            }
+        }
+
+        if (topicName.isPartitioned()) {
```

Comment:
same here, why invoking method for non-partitioned topic when `topicName.isPartitioned() `is `true`.

Reply:
maybe the method name is confused, what mean here is for single partition